---
title: "Day 6: Interpretating assembly results, AMR genes and Virulence Factors"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
---

```{r, include=FALSE}
library(knitr)
```

# Objectives
1. Inteprete assembly, annotation and quality assessment
2. Identify virulence factors
3. Identify AMR genes

# Day 5 recap
1. Running the genome assembly and annotation pipeline
2. Interprete the results and discuss their usefulness.
3. Identify virulence factors.
4. Identify AMR genes and MLST
5. Wrap-up.


# Bioinformatics pipeline for bacterial WGS

```{r explainining-permissions, echo=FALSE}
include_graphics("C:/Users/epi/Documents/bacterial_wgs_bioinformatics/bacterial_wgs_workflow.PNG")
```

# Assembly results and interpretation

```{r assembly-output, echo=FALSE}
include_graphics("C:/Users/epi/Documents/bacterial_wgs_bioinformatics/assembly_output.PNG")
```

 - One dir for each of the four included assemblers: Abyss, Velvet, SPades and IDBA, each with suffix Contigs, e.g., SPadesContigs.
 - Preprocessing dir - containing the reads after post-trim by Trimmomatic.
 - AssemblyReport - containing the QUAST (Quality Assessment Tool for Genome Assemblies) report.
 - IntegratedContigs - containing the master contigs formed by integrating the contigs generated by four assemblers.
 - AnnotatedContigs - containing master contigs annotated by PROKKA (Prokaryotic Genome Annotation Pipeline). PROKKA identifies features such as genes, coding sequences, rRNA, tRNA, and other genomic elements, providing functional annotations that are useful for downstream genomic analyses.
 
## Exercise
 - Create a folder and call it "assembly_preprocess".
 - Move all the contents of the "Preproccessing" dirs for all the species to this folder.
 - Check the quality of these reads using FASTQC and Multiqc.
 - Compare your results to those after quality filtering and trimming from the fastq_QC nextflow pipeline.
 
# Virulence factors
 - We determine the molecules produced by each species that contribute to its ability to cause diseases (virulence factors: VFs). 
 - These VFs enable the bacteria to colonize a host, evade or suppress the host's immune response, and obtain nutrients from the host.
 
## Why are virulence factors useful?

 - To understand Pathogenicity - such genes encode proteins that enhance a pathogenâ€™s ability to cause disease. Identifying these genes helps in understanding how a particular bacterium infects hosts and causes disease.
 - Knowledge of virulence factors can lead to the development of targeted treatments or vaccines that neutralize these factors.
 - To predict disease severity - The presence and expression levels of virulence genes can indicate the potential severity of an infection, aiding in clinical decision-making.
 - To track pathogens - Identifying virulence genes helps in tracking the spread and evolution of pathogenic strains in populations.
 
```{r virilulencefactors, eval=FALSE}
nano virulence_factors.sh
#!/bin/bash
#SBATCH --job-name='VFs'
#SBATCH --nodes=1 --ntasks=10
#SBATCH --time=48:00:00
#SBATCH --mem=80g
#SBATCH --output=/users/username/bacterial_wgs_training/logs/vfs-abricate-stdout.log
#SBATCH --error=/users/username/bacterial_wgs_training/logs/vfs-abricate-stderr.log
#SBATCH --mail-user=ephie.geza@uct.ac.za

set -e

module load python/3.12.1

proj="/users/username/bacterial_wgs_training/"
img="/cbio/training/courses/2024/bacterial_wgs/bacterial_wgs_latest-2024-07-22-32bd9a1b9ae4.simg"

for gen in Serratia Enterobacter Pseudomonas Escherichia
do
        output=${proj}"vfdb_abricate/"${gen}
        data=${proj}"assemblyout_"${gen}"/IntegratedContigs"

        if [ ! -d ${output} ]; then
                mkdir -p ${output}
        fi

        ## RUN abricate considering a minimum identity of 90% and minimum coverage of 90%
        cd ${data}
        for file in `ls *_master_integrated_contigs.fa`
        do
                singularity exec ${img} abricate --db vfdb \
                        ${data}/${file} --minid 90 --mincov 90  > ${output}/${file}_${gen}vfdb.tsv
        done
        # Summarize the results into one file
        singularity exec ${img} abricate --summary \
                ${output}/*_vfdb.tsv > ${output}/${gen}_summary_vfdb.tsv
done
```

# Antimicrobial resistance (AMR) genes

## Why is it important

 - To guide treatment - To informs clinicians about the antibiotics that are likely to be ineffective, guiding appropriate and effective treatment choices.
 - To monitor resistance trends - Surveillance of AMR genes helps in monitoring the emergence and spread of resistance in bacterial populations, crucial for public health strategies.
 - To prevent resistance spread - Understanding the genetic basis of resistance can help in devising strategies to prevent the spread of resistant strains, such as infection control measures and stewardship programs.
 - For research and development - can inform the development of new antibiotics and alternative treatments to overcome resistance.
 
```{r rgi-card, eval=FALSE}
#!/bin/bash
#SBATCH --job-name='rgi_final'
#SBATCH --nodes=1 --ntasks=10
#SBATCH --partition=Main
#SBATCH --mem=60GB
#SBATCH --output=/users/username/bacterial_wgs_training/logs/rgi_final-stdout.txt
#SBATCH --error=/users/username/bacterial_wgs_training/logs/rgi_final-stderr.txt
#SBATCH --time=12:00:00


# Set important dirs
proj="/users/username/bacterial_wgs_training/"
rgi_dir="/users/username/bacterial_wgs_training/02_AMR/rgi_final"
img="/cbio/projects/033/images/bacterial_wgs_latest-2024-07-22-32bd9a1b9ae4.simg"
db="/cbio/training/courses/2024/bacterial_wgs/localDB"
cd ${proj}
# Do for each species
cd /users/username/bacterial_wgs_training/assemblyout_${gen}
mkdir ${rgi_dir}

for i in `ls *.fna`
do
       #z=$(echo "$i" | cut -d'.' -f1)
       singularity exec ${img} rgi main --input_sequence ${i} --output_file ${rgi_dir}/${i} --local  --clean --include_loose
done

cd ${rgi_dir}
```

# Multi-locus Sequence Typing (MLST)

## Why MLST
 - Epidemiological tracking - provides a standardized method to identify and differentiate bacterial strains, facilitating the tracking of outbreaks and the study of bacterial epidemiology.
 - Population genetics - helps in understanding the genetic diversity, population structure, and evolutionary relationships of bacterial species.
 - Global comparisons - MLST data can be easily compared across different laboratories and geographic regions, aiding in global surveillance and collaborative research.
 - Identify clonal lineages - helps in identifying clonal lineages associated with specific outbreaks or resistance patterns, providing insights into the transmission dynamics of pathogenic bacteria.
 